File: backend/app/routes/analytics.py
from flask import Blueprint, jsonify, request
from app.models import db, Match, Player, Character, Stage
from sqlalchemy import func
from datetime import datetime, timedelta

analytics_bp = Blueprint('analytics', __name__)

@analytics_bp.route('/unique-stats', methods=['GET'])
def get_unique_stats():
    # Get unique counts
    unique_players = Player.query.count()
    unique_characters = Character.query.count()
    unique_stages = Stage.query.count()
    
    # Get date range
    first_match = Match.query.order_by(Match.datetime.asc()).first()
    last_match = Match.query.order_by(Match.datetime.desc()).first()
    
    date_range = {
        'start': first_match.datetime.isoformat() if first_match else None,
        'end': last_match.datetime.isoformat() if last_match else None
    }
    
    return jsonify({
        'unique_players': unique_players,
        'unique_characters': unique_characters,
        'unique_stages': unique_stages,
        'date_range': date_range
    })

@analytics_bp.route('/daily-stats', methods=['GET'])
def get_daily_stats():
    # Get daily statistics
    daily_stats = db.session.query(
        func.date(Match.datetime).label('date'),
        func.count(Match.id).label('matches'),
        func.count(func.distinct(Match.shayne_character_id)).label('unique_characters_shayne'),
        func.count(func.distinct(Match.matt_character_id)).label('unique_characters_matt'),
        func.count(func.distinct(Match.stage_id)).label('unique_stages')
    ).group_by(func.date(Match.datetime))\
     .order_by(func.date(Match.datetime).desc())\
     .all()
    
    return jsonify([{
        'date': stat.date.isoformat(),
        'matches': stat.matches,
        'unique_characters_shayne': stat.unique_characters_shayne,
        'unique_characters_matt': stat.unique_characters_matt,
        'unique_stages': stat.unique_stages
    } for stat in daily_stats])

@analytics_bp.route('/character-usage', methods=['GET'])
def get_character_usage():
    # Get character usage statistics
    character_stats = db.session.query(
        Character.name,
        func.count(Match.id).label('total_usage'),
        func.count(func.distinct(Match.shayne_id)).label('unique_players_shayne'),
        func.count(func.distinct(Match.matt_id)).label('unique_players_matt')
    ).outerjoin(Match, (Character.id == Match.shayne_character_id) | (Character.id == Match.matt_character_id))\
     .group_by(Character.name)\
     .order_by(func.count(Match.id).desc())\
     .all()
    
    return jsonify([{
        'character': stat.name,
        'total_usage': stat.total_usage,
        'unique_players_shayne': stat.unique_players_shayne,
        'unique_players_matt': stat.unique_players_matt
    } for stat in character_stats])

@analytics_bp.route('/stage-usage', methods=['GET'])
def get_stage_usage():
    # Get stage usage statistics
    stage_stats = db.session.query(
        Stage.name,
        func.count(Match.id).label('total_usage'),
        func.count(func.distinct(Match.shayne_character_id)).label('unique_characters_shayne'),
        func.count(func.distinct(Match.matt_character_id)).label('unique_characters_matt')
    ).outerjoin(Match, Stage.id == Match.stage_id)\
     .group_by(Stage.name)\
     .order_by(func.count(Match.id).desc())\
     .all()
    
    return jsonify([{
        'stage': stat.name,
        'total_usage': stat.total_usage,
        'unique_characters_shayne': stat.unique_characters_shayne,
        'unique_characters_matt': stat.unique_characters_matt
    } for stat in stage_stats])

@analytics_bp.route('/player-stats', methods=['GET'])
def get_player_stats():
    # Get player statistics
    player_stats = db.session.query(
        Player.name,
        func.count(Match.id).label('total_matches'),
        func.count(func.distinct(Match.shayne_character_id)).label('unique_characters'),
        func.count(func.distinct(Match.stage_id)).label('unique_stages')
    ).outerjoin(Match, (Player.id == Match.shayne_id) | (Player.id == Match.matt_id))\
     .group_by(Player.name)\
     .all()
    
    return jsonify([{
        'player': stat.name,
        'total_matches': stat.total_matches,
        'unique_characters': stat.unique_characters,
        'unique_stages': stat.unique_stages
    } for stat in player_stats]) 

File: frontend/src/pages/Analytics.js
import React, { useState, useEffect } from 'react';
import {
    Container,
    Grid,
    Paper,
    Typography,
    Box,
    CircularProgress,
    Alert,
    Card,
    CardContent,
    Divider,
} from '@mui/material';
import {
    LineChart,
    Line,
    XAxis,
    YAxis,
    CartesianGrid,
    Tooltip,
    Legend,
    ResponsiveContainer,
    BarChart,
    Bar,
} from 'recharts';
import axios from 'axios';

const Analytics = () => {
    const [loading, setLoading] = useState(true);
    const [error, setError] = useState(null);
    const [uniqueStats, setUniqueStats] = useState(null);
    const [dailyStats, setDailyStats] = useState([]);
    const [characterStats, setCharacterStats] = useState([]);
    const [stageStats, setStageStats] = useState([]);
    const [playerStats, setPlayerStats] = useState([]);

    useEffect(() => {
        const fetchData = async () => {
            try {
                setLoading(true);
                const [uniqueRes, dailyRes, charRes, stageRes, playerRes] = await Promise.all([
                    axios.get('/api/analytics/unique-stats'),
                    axios.get('/api/analytics/daily-stats'),
                    axios.get('/api/analytics/character-usage'),
                    axios.get('/api/analytics/stage-usage'),
                    axios.get('/api/analytics/player-stats')
                ]);

                setUniqueStats(uniqueRes.data);
                setDailyStats(dailyRes.data);
                setCharacterStats(charRes.data);
                setStageStats(stageRes.data);
                setPlayerStats(playerRes.data);
            } catch (err) {
                setError('Failed to fetch analytics data');
                console.error(err);
            } finally {
                setLoading(false);
            }
        };

        fetchData();
    }, []);

    if (loading) {
        return (
            <Box display="flex" justifyContent="center" alignItems="center" minHeight="80vh">
                <CircularProgress />
            </Box>
        );
    }

    if (error) {
        return (
            <Container>
                <Alert severity="error">{error}</Alert>
            </Container>
        );
    }

    return (
        <Container maxWidth="lg" sx={{ mt: 4, mb: 4 }}>
            <Typography variant="h4" gutterBottom>
                Analytics Dashboard
            </Typography>

            {/* Unique Statistics Cards */}
            <Grid container spacing={3} sx={{ mb: 4 }}>
                <Grid item xs={12} sm={6} md={3}>
                    <Paper sx={{ p: 2, textAlign: 'center' }}>
                        <Typography variant="h6">Unique Players</Typography>
                        <Typography variant="h4">{uniqueStats.unique_players}</Typography>
                    </Paper>
                </Grid>
                <Grid item xs={12} sm={6} md={3}>
                    <Paper sx={{ p: 2, textAlign: 'center' }}>
                        <Typography variant="h6">Unique Characters</Typography>
                        <Typography variant="h4">{uniqueStats.unique_characters}</Typography>
                    </Paper>
                </Grid>
                <Grid item xs={12} sm={6} md={3}>
                    <Paper sx={{ p: 2, textAlign: 'center' }}>
                        <Typography variant="h6">Unique Stages</Typography>
                        <Typography variant="h4">{uniqueStats.unique_stages}</Typography>
                    </Paper>
                </Grid>
                <Grid item xs={12} sm={6} md={3}>
                    <Paper sx={{ p: 2, textAlign: 'center' }}>
                        <Typography variant="h6">Date Range</Typography>
                        <Typography variant="body2">
                            {new Date(uniqueStats.date_range.start).toLocaleDateString()} -{' '}
                            {new Date(uniqueStats.date_range.end).toLocaleDateString()}
                        </Typography>
                    </Paper>
                </Grid>
            </Grid>

            {/* Daily Trends Chart */}
            <Paper sx={{ p: 2, mb: 4 }}>
                <Typography variant="h6" gutterBottom>
                    Daily Trends
                </Typography>
                <ResponsiveContainer width="100%" height={300}>
                    <LineChart data={dailyStats}>
                        <CartesianGrid strokeDasharray="3 3" />
                        <XAxis dataKey="date" />
                        <YAxis />
                        <Tooltip />
                        <Legend />
                        <Line type="monotone" dataKey="matches" stroke="#8884d8" name="Matches" />
                        <Line type="monotone" dataKey="unique_characters_shayne" stroke="#82ca9d" name="Unique Characters (Shayne)" />
                        <Line type="monotone" dataKey="unique_characters_matt" stroke="#ffc658" name="Unique Characters (Matt)" />
                        <Line type="monotone" dataKey="unique_stages" stroke="#ff7300" name="Unique Stages" />
                    </LineChart>
                </ResponsiveContainer>
            </Paper>

            {/* Character Usage Chart */}
            <Paper sx={{ p: 2, mb: 4 }}>
                <Typography variant="h6" gutterBottom>
                    Character Usage
                </Typography>
                <ResponsiveContainer width="100%" height={300}>
                    <BarChart data={characterStats}>
                        <CartesianGrid strokeDasharray="3 3" />
                        <XAxis dataKey="character" />
                        <YAxis />
                        <Tooltip />
                        <Legend />
                        <Bar dataKey="total_usage" fill="#8884d8" name="Total Usage" />
                        <Bar dataKey="unique_players_shayne" fill="#82ca9d" name="Unique Players (Shayne)" />
                        <Bar dataKey="unique_players_matt" fill="#ffc658" name="Unique Players (Matt)" />
                    </BarChart>
                </ResponsiveContainer>
            </Paper>

            {/* Stage Usage Chart */}
            <Paper sx={{ p: 2, mb: 4 }}>
                <Typography variant="h6" gutterBottom>
                    Stage Usage
                </Typography>
                <ResponsiveContainer width="100%" height={300}>
                    <BarChart data={stageStats}>
                        <CartesianGrid strokeDasharray="3 3" />
                        <XAxis dataKey="stage" />
                        <YAxis />
                        <Tooltip />
                        <Legend />
                        <Bar dataKey="total_usage" fill="#8884d8" name="Total Usage" />
                        <Bar dataKey="unique_characters_shayne" fill="#82ca9d" name="Unique Characters (Shayne)" />
                        <Bar dataKey="unique_characters_matt" fill="#ffc658" name="Unique Characters (Matt)" />
                    </BarChart>
                </ResponsiveContainer>
            </Paper>

            {/* Player Statistics */}
            <Grid container spacing={3}>
                {playerStats.map((player) => (
                    <Grid item xs={12} sm={6} key={player.player}>
                        <Paper sx={{ p: 2 }}>
                            <Typography variant="h6" gutterBottom>
                                {player.player}
                            </Typography>
                            <Typography>Total Matches: {player.total_matches}</Typography>
                            <Typography>Unique Characters: {player.unique_characters}</Typography>
                            <Typography>Unique Stages: {player.unique_stages}</Typography>
                        </Paper>
                    </Grid>
                ))}
            </Grid>
        </Container>
    );
};

export default Analytics; 